### ISSUE TITLE
fix new bug that errors out on models with ragged arrays

### ISSUE THREAD

**Issue by paciorek** [nimble-dev/nimble#752]
Code such as:
```
     Z[t, 1:n_obs_y[t]] <- K[t, 1:n_obs_y[t], 1:n_grid] %*% X[t, 1:n_grid]
```

is causing an error because in my attempt to generate NaNs when an invalid dynamic index is used, I needed to determine the dimension of the LHS of model declarations, e.g. that y[2,3:7] is one dimension with length 5. I used makeSizeAndDimList(), which was really only intended for use with AD, and it errors out when seeing a ragged array.

In having thought a bit more, I don't think we should ever have a ragged array on the LHS with dynamic indexing on the RHS, since that would mean that as an observation shifted from cluster  to cluster, it  could be using a different number of elements from the RHS var than another observation that was also in the same cluster, e.g.,
```
y[i, 1:n[i]] ~ dmnorm(mu[k[i], 1:n[i]], ....)
```
and suppose k[1]=1, k[2]=1, n[1]=3, n[2]=4.

I suppose one could imagine such models, but not clear they make sense. This suggests that we could continue to use makeSizeAndDimList but only call it if we know the BUGSdecl involves dynamic indexing. (Currently we are calling it for every declaration, unnecessarily, since the output of makeSizeAndDimList is only being used for expressions with dynamic indexing.) And then if ragged array is detected we'd error out. 

If this is a good path, we need:
1) to only call makeSizeAndDimList if dynamic indexing is turned on **and** if `BUGSdecl$dynamicIndexInfo` is not NULL
2) have the makeSizeAndDimList ragged array error message be more general to say that one can't use derivatives with ragged arrays (if derivs is enabled) and one can't have dynamically indexed ragged arrays.

Just laying out these initial thoughts. More discussion worthwhile.

**Comment by paciorek** [comment]
also, we clearly need a test that at least minimally uses a model with a ragged array!

**Comment by perrydv** [comment]
@paciorek I don't have the picture fully focused, but here is an idea.

We could make a version of `makeSizeAndDimList` that does not do the check for ragged arrays, assuming that is only relevant for AD. (Alternatively, we could add an argument to the current `makeSizeAndDimList` for whether to do the check.)  Would that solve the problem?

I am cautious about deciding that the combination you outlined could never be valid.  We've seen huge creativity in how people write models.  Unless we can prove it could never conceivably be valid, it would be nice to support it.

**Comment by paciorek** [comment]
Fixed by PR #764
